[WIP]RasterCache adds a new mechanism for particularly complex pictures or display lists

ColdPaleLight opened the pull request:
`RasterCach::Prepare` adds the new parameter `is_high_priority` to specify whether the raster cache entry is high priority. If the raster cache entry is high priority, it will always cache on first usage and survive 3 frames without usage.

Partially fix issues:
https://github.com/flutter/flutter/issues/87827 
https://github.com/flutter/flutter/issues/87826

If this PR can land, the following PRs will expose this mechanism to `Layer` of `dart:ui` and add zombie algorithm as mentioned in https://github.com/flutter/flutter/issues/87827#issuecomment-914537096

cc @flar @jonahwilliams 

## Pre-launch Checklist
- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

jonahwilliams commented:
In principle, I like this change, for the issues described in the bug. The only potential downside is that we cache more than we were caching before - but I think it is rare that the framework would display one 1 frame for something. most of the short animations are ~100ms at least.

jonahwilliams commented:
I think the bigger question is: if this change is worthwhile, then what is the downside of always caching on the first frame?

flar reviewed (changes requested):
We already have a flag for complexity and it is called "is_complex". We should not make a new flag that is "something something AND complex". We should keep the flags orthogonal. If something isn't complex then it is slower to cache it than it is to render it from scratch so providing yet another way for developers to defeat the complexity scoring is just another rope for them to hang themselves with.

ColdPaleLight commented:
> We already have a flag for complexity and it is called "is_complex". We should not make a new flag that is "something something AND complex". We should keep the flags orthogonal. If something isn't complex then it is slower to cache it than it is to render it from scratch so providing yet another way for developers to defeat the complexity scoring is just another rope for them to hang themselves with.

I was hesitant before whether to add a new flag. If `isComplex` is reused, many current tests will fail, because the picture or displaylist in the test depends on `isComplex` to participate in the raster cache. So we need to provide a new way to fix these tests. WDYT?

ColdPaleLight commented:
Hi, @flar 
I renamed the original `is_complex` to `for_testing`, provided it for unit testing. and renamed the flag `is_high_priority` to `is_complex`. This way we don't need to add new flags and we can make the unit tests work as before. Please let me know if this makes sense, thanks.

flar commented:
> > We already have a flag for complexity and it is called "is_complex". We should not make a new flag that is "something something AND complex". We should keep the flags orthogonal. If something isn't complex then it is slower to cache it than it is to render it from scratch so providing yet another way for developers to defeat the complexity scoring is just another rope for them to hang themselves with.
> 
> I was hesitant before whether to add a new flag. If `isComplex` is reused, many current tests will fail, because the picture or displaylist in the test depends on `isComplex` to participate raster cache. So if we reuse `isComplex`, we need to provide a new way to fix these tests. WDYT?

Oh dear, my comment was taken the wrong way.

My comment was about the *meaning* of the new flag, not whether or not to add one.

We have is_complex to mean the drawing is complex. You were adding "is_high_priority" and making it mean "is_complex" as well. I'm OK with 3 flags, but their meaning should be orthogonal. "is_high_priority" should only affect "when" we cache, not "whether it is complex enough to cache".

In particular, you had this code (in both the picture and display_list code):
```
  if (is_complex || is_high_priority) {
    // The caller seems to have extra information about the picture and thinks
    // the picture is always worth rasterizing.
    return true;
  }
```

The "high priority" flag should *not* be overriding our metrics as to whether or not the picture is worth rasterizing. If something is not worth rasterizing then we should not rasterize it *even if it is high priority*.

Granted, our "op_count > 5" metric is very rudimentary, but we have already implemented a better mechanism for the DisplayList code path that should be much more reliable in the future to the point that we might want to start ignoring the "is_complex" flag entirely, but we aren't there yet.

We are at a point, though, where they already have the ability to set "is_complex" to override that logic. We should not have another flag also indicate complexity. For whatever changes "high_priority" makes to our caching logic, the parts that it should not impact are the parts that are already controlled by "is_complex". Instead "high_priority" should impact whether we count down from 3 before we cache it and whether we keep it after its last use for a few frames, but it should not impact our complexity rating.

flar commented:
In short, I think what was intended here was 3 flags, meaning:

- will_change: **only** tells us that the picture will change frequently (on every frame)
- is_complex: **only** provides an override for our complexity metrics
- is_high_priority: **only** suggests we be more aggressive about populating the cache entry and more lax about evicting it

As your original PR was implemented, you had "is_high_priority" also interfering with our complexity evaluation - which is the job of "is_complex" and should not be a side effect of "is_high_priority".

ColdPaleLight commented:
@flar Sorry, I misunderstood your comment. Thank you for your kind comments, I tweaked the code as you suggested, please take a look again, thanks!

flar commented:
We recently revisited one of the issues that suggested this PR and were concerned that the needs were not adequately predetermined. In other words, this concept is a working hypothesis, not an immediate need.

I'm tagging @dnfield and @jonahwilliams to weigh in on the empirical evidence that this particular solution might address, or how to collect such evidence and evaluate that this solution makes a difference.

flar commented on flow/raster_cache.cc:
```diff
@@ -405,9 +412,25 @@ void RasterCache::SweepOneCacheAfterFrame(RasterCacheKey::Map<Entry>& cache,
+      if (entry.unused_count < entry.unused_threshold()) {
+        entry.unused_count++;
+        if (entry.image) {
+          RasterCacheKeyKind kind = it->first.kind();
```
Maybe have a function `metrics_for(kind)` or `metrics_for(<type of it.first>)` to simplify all 3 cases where we have to choose a metrics to modify?

ColdPaleLight replied:
done, added a new method `GetMetricsForKind`.

flar reviewed (changes requested):
Some minor nits and a potential logic change once we look at some more empirical data, but I think this concept is implemented well here.

I'm flagging this as  "Request changes" only on the basis that I'd like to see more buy-in on the concept and a way to measure this empirically before we move forward.

dnfield commented:
This has been tricky to nail down. I'm less convinced now that using the raster cache more will help us. Jonah may have more data on this, but in various experiments I tried using the raster cache more actually slowed things down. It'd help to see the benchmark(s) this helps.

ColdPaleLight commented on flow/raster_cache.h:
```diff
@@ -286,9 +301,16 @@ class RasterCache {
     std::unique_ptr<RasterCacheResult> image;
+    // Return the number of frames the entry survives if it is not used. If the
+    // number is 0, then it will be evicted when not in use.
+    size_t unused_threshold() const { return is_high_priority ? 3 : 0; }
```
done

dnfield reviewed (changes requested):
I think this is an interesting idea, but we need to have a better motivating case for this. Even in the linked case, we can see that some SVGs benefit from this and others regress. Our customers are likely to not use this correctly.

We already have two flags that feed into raster caching behavior, and that customers don't really know how to use and/or rarely use. 

What really need here are some solid benchmarks and docs that explain what we want to do with the raster cache. I personally have had a hard time coming up with motivating cases to show that using it more would be better, or that holding on to entries longer would be better. The GPU seems to start to get into a poor state when it has to work with too many textures, and making a change that will both be correct in most cases _and_ be relatively easy for developers to use is hard.

dnfield commented:
Happy to discuss this more on discord and/or in a doc.

ColdPaleLight commented:
1. Raster cache should allow a configuration to always cache on first usage
I've been trying for a long time and haven't found conclusive proof that this can benefit raster caching. In theory, this should pay off, but turning theory into reality looks very difficult.

2. Consider allowing raster cache entries to survive 1+ frames without usage
This link (https://github.com/flutter/flutter/issues/87827#issuecomment-1091846543) is the result of the benchmark. I believe that fixing this issue will benefit the scene where the scrollview scrolls up and down.

But it seems a bit difficult for the user to mark which pictures should not be cleaned up when unused. Maybe we should introduce an LRU cache to cache the unused picture raster cache entry.

flar commented:
> I investigated the two features that `is_high_priority` wanted to add separately.
> 
> 1. Raster cache should allow a configuration to always cache on first usage
>    I've been trying for a long time and haven't found conclusive proof that this can benefit raster caching. In theory, this should pay off, but turning theory into reality looks very difficult.
> 2. Consider allowing raster cache entries to survive 1+ frames without usage
>    This link ([Consider allowing raster cache entries to survive 1+ frames without usage flutter#87827 (comment)](https://github.com/flutter/flutter/issues/87827#issuecomment-1091846543)) is the result of the benchmark. I believe that fixing this issue will benefit the scene where the scrollview scrolls up and down.
>    But it seems a bit difficult for the user to mark which pictures should not be cleaned up when unused. Maybe we should introduce an LRU cache to cache the unused picture raster cache entry.

Doesn't flutter have a mechanism by which you can express a range around the visible field during which the elements will be included in the tree? If so, then they might remain cached if they are just off the screen. I think when I found it one day I asked around and there may be some issues with how it is implemented. If we can get that mechanism to work right, then the above conditions may fall out from it.

flar commented:
Other things to consider - does the extra memory for "potentially reusable" cache entries negatively impact apps? Does it cost extra batter to keep the app warm or to wake it up on a phone with a lot of background tasks?

ColdPaleLight commented:
> Doesn't flutter have a mechanism by which you can express a range around the visible field during which the elements will be included in the tree? If so, then they might remain cached if they are just off the screen. I think when I found it one day I asked around and there may be some issues with how it is implemented. If we can get that mechanism to work right, then the above conditions may fall out from it.

This is common in ListViews, if an item scrolls out of the visible area but is still in the cache area, it will be cached instead of destroyed.

c.f.
https://api.flutter.dev/flutter/widgets/ScrollView/cacheExtent.html
